### PR TITLE
add four mode for the element how to fit the htmlMesh size

### PR DESCRIPTION
**Pull request by CTOSPACE**
i added four mode to fit the HtmlMesh size, contain cover stretch none, i am not sure the mode 'cover' or 'stretch' would be useful, At least it won't be used in my current project, but I think it may be used in some situations, btw. i have removed setTimeout, it also get the right size of the element,  i will test it in my project later and find the reason.

### CODE REVIEW

File: HtmlMesh/src/html-mesh.ts [hunk 1]
```diff
@@ -6,6 +6,114 @@ import { PointerEventsCaptureBehavior } from "./pointer-events-capture-behavior"
...
 import { Logger } from "@babylonjs/core/Misc/logger";
 
+export enum SizeFit {
+    none = 1,
+    contain = 2,
+    cover = 3,
+    stretch = 4
+}
+
+export type SizeFitHandler = {
```
Comment (ericwood73) [inline HtmlMesh/src/html-mesh.ts:16]: How about FitStrategy with functions wrapElement and updateSize.  Let's make createElement without side effects so it accepts the element and returns the sizing wrapper.   updateSize would modify the passed in element.

File: HtmlMesh/src/html-mesh.ts [hunk 2]
```diff
@@ -6,6 +6,114 @@ import { PointerEventsCaptureBehavior } from "./pointer-events-capture-behavior"
...
+    handleElement(element: HTMLElement): HTMLElement
+    handleSizeFit(parent: HTMLElement, width: number, height: number): void
```
Comment (ericwood73) [inline HtmlMesh/src/html-mesh.ts:18]: The implementations never seem to use parent element so why don't we just make the parameter the sizing element (the element returned by wrapElement).

File: HtmlMesh/src/html-mesh.ts [hunk 3]
```diff
@@ -6,6 +6,114 @@ import { PointerEventsCaptureBehavior } from "./pointer-events-capture-behavior"
...
+    handleSizeFit(parent: HTMLElement, width: number, height: number) {
+        const childElement = parent.firstElementChild as HTMLElement;
+        if (childElement) {
+            childElement.style.width = `${width}px`;
+            childElement.style.height = `${height}px`;
+        }
+    }
+}
+
+const SizeFitHandlerMap = {
```
Comment (ericwood73) [inline HtmlMesh/src/html-mesh.ts:110]: Can we move all of the FitStrategy stuff into its own ts file.  Then we would export the FitStrategy type as well as FitStrategy.CONTAIN, FitStrategy.COVER, etc...  (Similar to how the RenderingGroup compare functions are exposed).

File: HtmlMesh/src/html-mesh.ts [hunk 4]
```diff
@@ -64,7 +174,7 @@ export class HtmlMesh extends Mesh {
     constructor(
         scene: Scene,
         id: string,
-        { captureOnPointerEnter = true, isCanvasOverlay = false } = {}
+        { captureOnPointerEnter = true, isCanvasOverlay = false, sizeFit = SizeFit.none } = {}
```
Comment (ericwood73) [inline HtmlMesh/src/html-mesh.ts:177]: fitStrategy = FitStrategy.NONE

File: HtmlMesh/src/html-mesh.ts [hunk 5]
```diff
@@ -76,6 +186,7 @@ export class HtmlMesh extends Mesh {
             return;
         }
 
+        this._sizeFit = sizeFit
```
Comment (ericwood73) [inline HtmlMesh/src/html-mesh.ts:189]: this._fitStrategy = fitStrategy

File: HtmlMesh/src/html-mesh.ts [hunk 6]
```diff
@@ -166,7 +278,8 @@ export class HtmlMesh extends Mesh {
         this.scaling.setAll(1);
 
         if (element) {
-            this._element!.appendChild(element);
+            const sizeFitHandler = SizeFitHandlerMap[this._sizeFit]
```
Comment (ericwood73) [inline HtmlMesh/src/html-mesh.ts:281]: This all gets replaced with `this._element!.appendChild(this._fitStrategy.wrapElement(element)`

File: HtmlMesh/src/html-mesh.ts [hunk 7]
```diff
@@ -195,11 +308,9 @@ export class HtmlMesh extends Mesh {
             return;
         }
 
-        const childElement = this._element!.firstElementChild as HTMLElement;
-        if (childElement) {
-            childElement.style.width = `${width}px`;
-            childElement.style.height = `${height}px`;
-        }
+        const sizeFitHandler = SizeFitHandlerMap[this._sizeFit]
```
Comment (ericwood73) [inline HtmlMesh/src/html-mesh.ts:311]: This all gets replaced with `this._fitStrategy.updateSize(this._element, width, height)`

**Review by ericwood73 — COMMENTED** [review]
Looks pretty good.  Would just like to change to follow a strategy pattern and move to it's own TS file (see my inline comments for specifics).  Also could you add content to both READMEs to describe the fit modes and how they differ.

File: HtmlMesh/src/html-mesh.ts [hunk 8]
```diff
@@ -6,6 +6,114 @@ import { PointerEventsCaptureBehavior } from "./pointer-events-capture-behavior"
...
+    cover = 3,
+    stretch = 4
+}
+
+export type SizeFitHandler = {
+    handleElement(element: HTMLElement): HTMLElement
+    handleSizeFit(parent: HTMLElement, width: number, height: number): void
+}
+
+const SizeFitContainHandler: SizeFitHandler = {
```
Comment (ericwood73) [inline HtmlMesh/src/html-mesh.ts:21]: FitStrategy.CONTAIN = {

File: HtmlMesh/src/html-mesh.ts [hunk 9]
```diff
@@ -6,6 +6,114 @@ import { PointerEventsCaptureBehavior } from "./pointer-events-capture-behavior"
...
+        sizingElement.style.height = `${height}px`
+
+        const [childWidth, childHeight] = [scalingElement!.offsetWidth, scalingElement!.offsetHeight]
+        const scale = Math.min(width / childWidth, height / childHeight)
+        scalingElement.style.transform = `scale(${scale})`
+        scalingElement.style.visibility = 'visible'
+    }
+}
+
+const SizeFitCoverHandler: SizeFitHandler = {
```
Comment (ericwood73) [inline HtmlMesh/src/html-mesh.ts:46]: FitStrategy.COVER = {

File: HtmlMesh/src/html-mesh.ts [hunk 10]
```diff
@@ -6,6 +6,114 @@ import { PointerEventsCaptureBehavior } from "./pointer-events-capture-behavior"
...
+
+        const [childWidth, childHeight] = [scalingElement!.offsetWidth, scalingElement!.offsetHeight]
+        const scale = Math.max(width / childWidth, height / childHeight)
+        scalingElement.style.transform = `scale(${scale})`
+        scalingElement.style.visibility = 'visible'
+    }
+}
+
+
+const SizeFitStretchHandler: SizeFitHandler = {
```
Comment (ericwood73) [inline HtmlMesh/src/html-mesh.ts:73]: FitStrategy.STRETCH = {

File: HtmlMesh/src/html-mesh.ts [hunk 11]
```diff
@@ -6,6 +6,114 @@ import { PointerEventsCaptureBehavior } from "./pointer-events-capture-behavior"
...
+        sizingElement.style.width = `${width}px`
+        sizingElement.style.height = `${height}px`
+
+        const [childWidth, childHeight] = [scalingElement!.offsetWidth, scalingElement!.offsetHeight]
+        scalingElement.style.transform = `scale(${width / childWidth}, ${height / childHeight})`
+        scalingElement.style.visibility = 'visible'
+    }
+}
+
+const SizeFitNoneHandler: SizeFitHandler = {
```
Comment (ericwood73) [inline HtmlMesh/src/html-mesh.ts:97]: FitStrategy.NONE =

File: HtmlMesh/src/html-mesh.ts [hunk 12]
```diff
@@ -6,6 +6,114 @@ import { PointerEventsCaptureBehavior } from "./pointer-events-capture-behavior"
 import { Scene } from "@babylonjs/core";
 import { Logger } from "@babylonjs/core/Misc/logger";
 
+export enum SizeFit {
```
[the remark this answers is not in the document]
Comment (CTOSPACE) [reply HtmlMesh/src/html-mesh.ts:9]: I have viewed your feedback, indeed, modifying the code according to your approach would make it looks better and more appropriate, i'll make the modifications and repush pr

**Pull request closed without merging**